misc/drmips.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2013-2015 Bruno Nova,
.\"
.TH DRMIPS 1 "September 14, 2015"
.SH NAME
drmips \- educational MIPS simulator
.SH SYNOPSIS
.B drmips
.RI [ OPTIONS ]
.RI [ FILE ]
.br
.SH DESCRIPTION
.B DrMIPS
is a graphical simulator of the MIPS processor to support computer architecture
teaching and learning. It is intuitive, versatile and configurable.
The simulator is available not only for personal computers but also for
Android devices, especially tablets.
.PP
This command starts the PC version of the simulator. You can pass a
.IR FILE
as argument to open it in the simulator.
.SH OPTIONS
.TP
.B \-\-opengl
Enable OpenGL hardware acceleration.
.TP
.B \-\-no\-opengl
Disable OpenGL hardware acceleration.
.TP
.B \-h, \-\-help
Display help and exit.
.TP
.B \-\-version
Display version information and exit.
.TP
.B FILE
Optional file to open.
.SH SEE ALSO
The manuals of the simulator at:
.IR /usr/share/doc/drmips/manuals/
.PP
The page of the simulator at:
.IR http://brunonova.github.io/drmips/


NEWS
v2.0.3
------

* Added Brazilian Portuguese translation to the PC version of the simulator
  (thanks crcpucmg).


v2.0.2
------

General:
* Fixed Bug #5, concerning a forwarding issue in the Register Bank
  (thanks Macmod)

Android:
* Target API version 25


v2.0.1
------

* Rewrote the manuals in HTML, dropping the build dependency on LaTeX.


v2.0.0
------

General:
* Custom CPU components can now be provided along with the CPU files and used
  by them
* The simulator now requires Java version 7 (up from version 6)
* The Java package names of the simulator where simplified

Android:
* Minimum required Android version increased to 4.0
* Dialogs are now built using Fragments, which should make them more robust

Note:
The Java packages of the simulator were renamed.
Because of this, the new version will be started with the default settings.
Additionally, the new version of the Android application will be installed
alongside the old version, instead of upgrading, so you will have to uninstall
the old version manually. Don't forget to backup your custom CPU and code files
before uninstalling!

Also note the names of the CPU components where changed (i.e. the values of
the "type" fields are different).


v1.3.2
------

* Moved the project to GitHub


v1.3.1
------

PC version:
* Disabled OpenGL hardware acceleration by default, and added an option to
  enable it (Bug BB-20)
* Switched to JOpt Simple to parse command-line arguments
* Added the "--reset" command-line argument that resets all settings to their
  defaults
* Handle uncaught exceptions properly


v1.3.0
------

General:
* Improved the readability and color scheme of the simulator (Bug BB-17)
* The assembled code table is now scrolled automatically to the instruction(s)
  currently being executed
* The inputs/outputs that are in the critical path are now highlighted in the
  tooltips of the components while in performance mode
* Added an option to display the names in the input/output tips
* Added an option to display the input/output tips for all components
* The input/output tips are now visible in performance mode as well

PC version:
* Added zoom controls to the datapath, including an auto-resize
  option (Bug BB-9)
* Improved the "internal windows" mode by using jScroll. It now has
  scrollbars, a toolbar to select windows, and options to tile and cascade
  windows (Bug BB-18)
* On exit, the simulator will now remember whether the window was maximized
  and its non-maximized size
* Added an option to show an 80 columns margin line in the code editor
* Added a status bar to the code editor that displays the caret's position
* Added an option to open the last file at startup
* The "print" feature now prints with the correct colors (Bug BB-3)
* Fixed wrong title being displayed in GNOME 3 (Bug BB-15). The fix doesn't
  work in older GNOME versions
* OpenGL hardware acceleration enabled, which should improve the performace,
  particularly in Linux

Android version:
* Converted the Android project to Android Studio


v1.2.3
------

PC version:
* JTattoo, RSyntaxTextArea and Autocomplete libraries are now compiled from
  source and included inside of the JAR file
* Updated JTattoo library

Android version:
* Removed a library that was not being used


v1.2.2
------

* Support Java 6


v1.2.1
------

* Added Portuguese translation of the configuration manual
* Improved the descriptions of the components
* Renamed the identifiers of the components and inputs
* Replaced JSON.org library with a free version


v1.2.0
------

* Fixed a bug where the calculation of the critical path would fail when the
  latencies of all components were 0
* Separated the cross-platform simulator code from the PC and Android projects
  into a new project
* Some refactoring


v1.1.1
------

General:
* Added credits to the "about" dialogs
* Switched to CMake build system to compile the simulator from the command-line

PC version:
* Added some command-line arguments to the PC version


v1.1.0
------

* The critical path of the current instruction can now be displayed.
  The global critical path of the CPU is still available


v1.0.6
------

PC version:
* Fixed problems with Windows network (UNC) paths (Bug BB-1)


v1.0.5
------

* All inputs that are only used at the end of the clock cycle now correctly
  don't affect the component's latency


v1.0.4
------

* Workaround for wrong latencies in Register Bank


v1.0.3
------

General:
* The tooltip of the ALU in the datapath now displays the operation it's
  performing
* Updated the latencies of the components to more "realistic" values

PC version:
* Added a user manual in English and Portuguese
* Corrected some translations


v1.0.2
------

General:
* Added an option to view some execution statistics
* Added an option to remove the latencies of all components (set them to zero)

PC version:
* Added a button to view the list of instructions supported by the CPU
* Improved the internationalization feature
* Updated the used libraries


v1.0.1
------

First open-source release


src/pc/src/main/java/brunonova/drmips/pc/DrMIPS.java
/*
    DrMIPS - Educational MIPS simulator
    Copyright (C) 2013-2015 Bruno Nova

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package brunonova.drmips.pc;

import brunonova.drmips.simulator.AppInfo;
import java.io.File;
import java.net.URI;
import java.util.Arrays;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.prefs.Preferences;
import javax.swing.JOptionPane;
import javax.swing.SwingUtilities;
import javax.swing.ToolTipManager;
import joptsimple.OptionParser;
import joptsimple.OptionSet;
import joptsimple.OptionSpec;


/**
 * Entry point of the program.
 * <p>It also contains some global constants and parameters.</p>
 *
 * @author Bruno Nova
 */
public class DrMIPS {
	/** The number of milliseconds before tooltips are displayed. */
	public static final int TOOLTIP_INITIAL_DELAY = 200;
	/** The number of milliseconds tooltips are shown before disappearing. */
	public static final int TOOLTIP_DISMISS_DELAY = 30000;
	/** Reference to the user preferences. */
	public static Preferences prefs = Preferences.userNodeForPackage(DrMIPS.class);
	/** The full path to the program's jar folder, or '.' if not running from a jar (when running from the IDE, for example). */
	public static String path = ".";
	/** The CPU file loaded by default. */
	public static final String DEFAULT_CPU = "cpu" + File.separator + "unicycle.cpu";
	/** Relative path to the documentation directory. */
	public static final String DOC_DIR = "doc";
	/** Absolute path to the fallback documentation directory (when installed in a separate directory). */
	public static final String DOC_DIR2 = "/usr/share/doc/drmips/manuals";

	// Names of the preferences
	public static final String LANG_PREF = "lang";
	public static final String LAST_CPU_PREF = "last_cpu";
	public static final String RECENT_FILES_PREF = "recent";
	public static final int MAX_RECENT_FILES = 10;
	public static final String RECENT_CPUS_PREF = "recent_cpu";
	public static final int MAX_RECENT_CPUS = 10;
	public static final String REGISTER_FORMAT_PREF = "reg_format";
	public static final String DATAPATH_DATA_FORMAT_PREF = "datapath_format";
	public static final String ASSEMBLED_CODE_FORMAT_PREF = "assembled_code_format";
	public static final String DATA_MEMORY_FORMAT_PREF = "data_memory_format";
	public static final String CODE_TAB_SIDE_PREF = "code_tab_side";
	public static final String DATAPATH_TAB_SIDE_PREF = "datapath_tab_side";
	public static final String REGISTERS_TAB_SIDE_PREF = "reg_tab_side";
	public static final String ASSEMBLED_CODE_TAB_SIDE_PREF = "assembled_code_tab_side";
	public static final String DATA_MEMORY_TAB_SIDE_PREF = "data_mem_tab_side";
	public static final String DIVIDER_LOCATION_PREF = "div_location";
	public static final String ASSEMBLE_RESET_PREF = "assemble_reset";
	public static final String DARK_THEME_PREF = "dark_theme";
	public static final String INTERNAL_WINDOWS_PREF = "internal_windows";
	public static final String SHOW_CONTROL_PATH_PREF = "show_control_path";
	public static final String SHOW_ARROWS_PREF = "show_arrows";
	public static final String PERFORMANCE_MODE_PREF = "performance_mode";
	public static final String PERFORMANCE_TYPE_PREF = "performance_type";
	public static final String OVERLAYED_DATA_PREF = "overlayed_data";
	public static final String OVERLAYED_SHOW_NAMES_PREF = "overlayed_show_names";
	public static final String OVERLAYED_SHOW_FOR_ALL_PREF = "overlayed_show_for_all";
	public static final String MAXIMIZED_PREF = "maximized";
	public static final String WIDTH_PREF = "width";
	public static final String HEIGHT_PREF = "height";
	public static final String MARGIN_LINE_PREF = "margin_line";
	public static final String LAST_FILE_PREF = "last_file";
	public static final String OPEN_LAST_FILE_AT_STARTUP_PREF = "open_last_file_at_startup";
	public static final String SCALE_PREF = "scale";
	public static final String AUTO_SCALE_PREF = "auto_scale";
	public static final String OPENGL_PREF = "use_opengl";

	// Default values of the preferences
	public static final int DEFAULT_REGISTER_FORMAT = Util.DECIMAL_FORMAT_INDEX;
	public static final int DEFAULT_DATAPATH_DATA_FORMAT = Util.DECIMAL_FORMAT_INDEX;
	public static final int DEFAULT_ASSEMBLED_CODE_FORMAT = Util.DECIMAL_FORMAT_INDEX;
	public static final int DEFAULT_DATA_MEMORY_FORMAT = Util.DECIMAL_FORMAT_INDEX;
	public static final int DEFAULT_PERFORMANCE_TYPE = Util.CPU_PERFORMANCE_TYPE_INDEX;
	public static final int DEFAULT_CODE_TAB_SIDE = Util.LEFT;
	public static final int DEFAULT_DATAPATH_TAB_SIDE = Util.LEFT;
	public static final int DEFAULT_REGISTERS_TAB_SIDE = Util.RIGHT;
	public static final int DEFAULT_ASSEMBLED_CODE_TAB_SIDE = Util.LEFT;
	public static final int DEFAULT_DATA_MEMORY_TAB_SIDE = Util.LEFT;
	public static final boolean DEFAULT_ASSEMBLE_RESET = true;
	public static final boolean DEFAULT_DARK_THEME = false;
	public static final boolean DEFAULT_INTERNAL_WINDOWS = false;
	public static final boolean DEFAULT_SHOW_CONTROL_PATH = true;
	public static final boolean DEFAULT_SHOW_ARROWS = true;
	public static final boolean DEFAULT_PERFORMANCE_MODE = false;
	public static final boolean DEFAULT_OVERLAYED_DATA = true;
	public static final boolean DEFAULT_OVERLAYED_SHOW_NAMES = false;
	public static final boolean DEFAULT_OVERLAYED_SHOW_FOR_ALL = false;
	public static final boolean DEFAULT_MAXIMIZED = true;
	public static final int DEFAULT_WIDTH = 800;
	public static final int DEFAULT_HEIGHT = 600;
	public static final boolean DEFAULT_MARGIN_LINE = false;
	public static final boolean DEFAULT_OPEN_LAST_FILE_AT_STARTUP = false;
	public static final double DEFAULT_SCALE = 1.0;
	public static final boolean DEFAULT_AUTO_SCALE = false;
	public static final boolean DEFAULT_OPENGL = false;

	/** "Loading" dialog. */
	private static DlgLoading dlgLoading = null;
	/** Main window. */
	private static FrmSimulator frmSim = null;
	/** Optional filename to open. */
	private static String filename = null;
	/** Class logger. */
	private static final Logger LOG = Logger.getLogger(DrMIPS.class.getName());

	private static void displayHelpAndExit(OptionParser parser) {
		// The help text will contain a "Usage" line. The default value of the line is:
		//   Usage: java -jar DrMIPS.jar [options] [file]
		// If the simulator will be started from a launcher script/program, you may
		// want to change the displayed "Usage" line. To do that, you must define
		// the "program.name" property in the launcher script, like so:
		//   java -Dprogram.name=drmips -jar /path/to/DrMIPS.jar
		// This example will change the "Usage" line to:
		//   Usage: drmips [options] [file]
		String prog_name = System.getProperty("program.name", "java -jar " + AppInfo.NAME + ".jar");

		System.out.println(AppInfo.NAME + " - " + AppInfo.DESCRIPTION);
		System.out.println("Usage: " + prog_name + " [options] [file]\n");
		try {
			parser.printHelpOn(System.out);
		} catch (Exception ex) {
			LOG.log(Level.SEVERE, "failed to print help", ex);
		}
		System.exit(0);
	}

	private static void displayVersionAndExit() {
		System.out.println(AppInfo.NAME + " " + AppInfo.VERSION + "\n"
			+ AppInfo.COPYRIGHT + "\n"
			+ "License: " + AppInfo.LICENSE_SHORT);
		System.exit(0);
	}

	private static void enableOpenGl() {
		// Hardware acceleration using OpenGL
		System.setProperty("sun.java2d.opengl", "True");
	}

	@SuppressWarnings("UseSpecificCatch")
	public static void main(String[] args) {
		boolean useOpenGl = prefs.getBoolean(OPENGL_PREF, DEFAULT_OPENGL);

		// Setup the default exception handler
		Thread.setDefaultUncaughtExceptionHandler(new ExceptionHandler());

		// Parse command-line arguments
		try {
			OptionParser parser = new OptionParser();
			OptionSpec<String> fileArg = parser.nonOptions("code file to open")
											   .ofType(String.class).describedAs("file");
			parser.acceptsAll(Arrays.asList("h", "help"), "display this help and exit").forHelp();
			parser.accepts("version", "display version information and exit");
			parser.accepts("opengl", "enable OpenGL hardware acceleration");
			parser.accepts("no-opengl", "disable OpenGL hardware acceleration");
			parser.accepts("reset", "reset all settings to their defaults");

			OptionSet options = parser.parse(args);
			List<String> otherArgs = options.valuesOf(fileArg);
			if(options.has("help"))
				displayHelpAndExit(parser);
			else if(options.has("version"))
				displayVersionAndExit();
			if(options.has("reset")) {
				prefs.clear();
				useOpenGl = DEFAULT_OPENGL;
			}
			if(options.has("no-opengl")) {
				useOpenGl = false;
				prefs.putBoolean(OPENGL_PREF, useOpenGl);
			} else if(options.has("opengl")) {
				useOpenGl = true;
				prefs.putBoolean(OPENGL_PREF, useOpenGl);
			}
			if(!otherArgs.isEmpty()) {
				if(otherArgs.size() == 1)
					filename = otherArgs.get(0);
				else {
					System.err.println("Only one file name should be supplied!");
					System.exit(1);
				}
			}
		} catch(Exception ex) {
			System.err.println("Error parsing arguments: " + ex.getMessage());
			System.exit(1);
		}

		// Try to enable OpenGL hardware acceleration, unless requested not to
		if(useOpenGl)
			enableOpenGl();

		// Find the path to the program
		try {
			URI uri = DrMIPS.class.getProtectionDomain().getCodeSource().getLocation().toURI(); // get the path to the jar (if running from a jar)
			String p = uri.toString();
			p = p.startsWith("file://") ? p.substring(5) : uri.getPath(); // uri to String (careful with Windows network (UNC) paths)
			if(p.toLowerCase().endsWith(".jar")) { // if running from a jar, get the path of the parent dir
				File f = (new File(p)).getParentFile();
				if(f != null) path = f.getCanonicalPath();
				LOG.log(Level.INFO, "running from jar file; path to jar folder: {0}", path);
			}
			else
				LOG.info("not running from jar file");
		} catch (Exception ex) {
			LOG.log(Level.WARNING, "error finding the path of the program", ex);
		}

		// Start the GUI
		SwingUtilities.invokeLater(new StartGUIRunnable());
	}

	private static class StartGUIRunnable implements Runnable {
		@Override
		public void run() {
			// Change tooltip delays
			ToolTipManager.sharedInstance().setInitialDelay(TOOLTIP_INITIAL_DELAY);
			ToolTipManager.sharedInstance().setDismissDelay(TOOLTIP_DISMISS_DELAY);

			// Display a "loading" dialog
			dlgLoading = new DlgLoading();
			dlgLoading.setVisible(true);
			Util.centerWindow(dlgLoading);
			dlgLoading.paint(dlgLoading.getGraphics());

			// Set the theme
			if(prefs.getBoolean(DARK_THEME_PREF, DEFAULT_DARK_THEME))
				Util.setDarkLookAndFeel();
			else
				Util.setLightLookAndFeel();
			SwingUtilities.updateComponentTreeUI(dlgLoading);

			// Load the strings
			if(!Lang.loadPreferredLanguage()) {
				JOptionPane.showMessageDialog(null, "Error opening language file " + Lang.getLanguage() + "!", AppInfo.NAME, JOptionPane.ERROR_MESSAGE);
				System.exit(1);
			}

			// Translate the "loading" window
			dlgLoading.setLoadingText(Lang.t("loading"));
			dlgLoading.paint(dlgLoading.getGraphics());

			// Start the main window
			if(filename == null && prefs.getBoolean(OPEN_LAST_FILE_AT_STARTUP_PREF, DEFAULT_OPEN_LAST_FILE_AT_STARTUP)) {
				String f = prefs.get(LAST_FILE_PREF, null);
				if(f != null && new File(f).isFile())
					filename = f;
			}
			frmSim = (filename == null ? new FrmSimulator() : new FrmSimulator(filename));
			frmSim.setVisible(true);

			dlgLoading.dispose();
		}
	}

	private static class ExceptionHandler implements Thread.UncaughtExceptionHandler {
		@Override
		public void uncaughtException(Thread t, Throwable ex) {
			LOG.log(Level.SEVERE, "error starting simulator", ex);
			JOptionPane.showMessageDialog(null, "Fatal error!\n" + ex, AppInfo.NAME, JOptionPane.ERROR_MESSAGE);
			if(frmSim == null || !frmSim.isVisible())
				System.exit(1);
		}
	}
}
